### PR TITLE
Support read device info when ecc is on

### PR DESCRIPTION
**Pull request by antrix1989**
## Proposed changes:
- Support read device info when ecc is on.
-  Unify the api (return same response as SSO ext returns for primary registration call)

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [x] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [x] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: IdentityCore/IdentityCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -8403,12 +8403,15 @@
 			baseConfigurationReference = D6CF4E961FC3626A00CD70C5 /* identitycore__tests__ios.xcconfig */;
 			buildSettings = {
 				CLANG_ENABLE_MODULES = YES;
+				CODE_SIGN_IDENTITY = "Apple Development";
```
Comment (oldalton) [inline IdentityCore/IdentityCore.xcodeproj/project.pbxproj:8406]: I think those changes are unintentional?

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m [hunk 2]
```diff
@@ -390,5 +453,47 @@ + (NSString *)getPrimaryEccTenantWithSharedAccessGroup:(NSString *)sharedAccessG
...
+    
+    NSError *subError = nil;
+    NSDictionary *decodedDataDict = [NSJSONSerialization JSONObjectWithData:metadataBlob
+                                                                    options:0
+                                                                      error:&subError];
+    if (!decodedDataDict || subError)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDKeychainErrorDomain, status, @"WPJ metadata deserialization failed.", nil, nil, subError, context.correlationId, nil, NO);
```
Comment (oldalton) [inline IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:484]: Should this error be logged?

**Review by oldalton — APPROVED** [review]

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m [hunk 3]
```diff
@@ -36,6 +37,43 @@
...
+    {
+        result[MSID_PRIMARY_REGISTRATION_CERTIFICATE_THUMBPRINT] = self.certificateThumbprint;
+        result[MSID_PRIMARY_REGISTRATION_CLOUD] = self.cloudHost;
+        result[MSID_PRIMARY_REGISTRATION_DEVICE_ID] = self.deviceID;
+        result[MSID_PRIMARY_REGISTRATION_TENANT_ID] = self.tenantIdentifier;
+        result[MSID_PRIMARY_REGISTRATION_UPN] = self.upn;
+    }
+    else
+    {
+        result[MSID_DEVICE_INFORMATION_AAD_DEVICE_ID_KEY] = self.deviceID;
```
Comment (Veena11) [inline IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:67]: for ecc registration, we dont send thumbprint and cloudHost ?

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m [hunk 4]
```diff
@@ -390,5 +453,47 @@ + (NSString *)getPrimaryEccTenantWithSharedAccessGroup:(NSString *)sharedAccessG
...
+    query[(id)kSecReturnData] = (id)kCFBooleanTrue;
+    
+    CFDictionaryRef attributeDictCF = NULL;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query,(CFTypeRef *)&attributeDictCF);
+    
+    NSDictionary *attributeDictionary = CFBridgingRelease(attributeDictCF);
+    NSData *metadataBlob = [attributeDictionary objectForKey:(__bridge id)kSecValueData];
+    
+    NSError *subError = nil;
+    NSDictionary *decodedDataDict = [NSJSONSerialization JSONObjectWithData:metadataBlob
```
Comment (Veena11) [inline IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:477]: should we use deserialize api from MSIDJsonSerializer ?

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:67]: For this case we send only 3 params (deviceID, upn, tenant). This format is being used by Defender iOS and they don't need other fields.  Also this is only when we read metadata locally, broker doesn't know about this "local" format.

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m (hunk 4 above)
Comment (antrix1989) [reply IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m]: Good point, addressed.

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m]: done

File: IdentityCore/IdentityCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/IdentityCore.xcodeproj/project.pbxproj]: yep, removed

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m (hunk 3 above)
Comment (Veena11) [reply IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:68]: aah ok, did we close on Defender's dependency on certificateThumbprint ?. They were using it for comparing against the certificate thumbprint in graph.

**Review by Veena11 — APPROVED** [review]

File: IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:68]: as far as I get, MacOS team needs thumbprint when iOS team doesn't. iOS defender uses this api in MSAL:
```

- (void)getWPJMetaDataDeviceWithParameters:(nullable MSALParameters *)parameters
                               forTenantId:(nullable NSString *)tenantId
                           completionBlock:(nonnull MSALWPJMetaDataCompletionBlock)completionBlock;
```

which returns in the local format and doesn't send request to the broker.

Defender for MacOS will be using this api:

```
- (void)getDeviceInformationWithParameters:(nullable MSALParameters *)parameters
                           completionBlock:(nonnull MSALDeviceInformationCompletionBlock)completionBlock;
```

it sends the request to broker and we return all data (including thumbprint) and the response in "primary/broker" format.

Comment (Veena11) [reply IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m:68]: aah sorry i missed this was only for ios. Thanks for explaining Serhii

**Pull request merged**